File: examples/js/product-consumer/verify-provider.test.ts
import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import { resolve } from 'node:path';
import { MessageProviderPact, Verifier } from '@pact-foundation/pact';

import { buildInventoryChangedEvent, startProviderServer } from './provider-server';

process.env.PACT_GRAPHQL_PLUGIN_VERSION ??= '0.1.0';

describe('GraphQL provider verification', () => {
  let server: Awaited<ReturnType<typeof startProviderServer>>;

  beforeAll(async () => {
    server = await startProviderServer();
  });

  afterAll(async () => {
    await server.close();
  });

  it('builds inventory message payload from provider logic', () => {
    const event = buildInventoryChangedEvent('var-1');
    expect(event).toEqual({
      subscription: 'InventoryChanged',
      variables: { variantId: 'var-1' },
      data: {
        inventoryChanged: { quantity: 42, updatedAt: '2026-03-08T12:00:00Z' },
      },
    });
  });

  it('verifies HTTP and message pacts', async () => {
    const httpPactPath = resolve(
      __dirname,
      'pacts',
      'product-consumer-product-provider.json',
    );
    const messagePactPath = resolve(
      __dirname,
      'pacts',
      'messages',
      'product-consumer-product-provider.json',
    );

    const httpVerifier = new Verifier({
      provider: 'product-provider',
      providerBaseUrl: server.url,
      pactUrls: [httpPactPath],
    });

    const messagePact = new MessageProviderPact({
      provider: 'product-provider',
      pactUrls: [messagePactPath],
      messageProviders: {
        'a GraphQL subscription event': async () =>
          buildInventoryChangedEvent('var-1'),
      },
    });

    await httpVerifier.verifyProvider();
    await messagePact.verify();
  });
});

File: examples/go/product-consumer/verify_provider_test.go
//go:build provider
// +build provider

package productconsumer

import (
	"context"
	"path/filepath"
	"testing"

	"github.com/pact-foundation/pact-go/v2/message"
	"github.com/pact-foundation/pact-go/v2/models"
	"github.com/pact-foundation/pact-go/v2/provider"
)

func TestBuildInventoryChangedEvent(t *testing.T) {
	event, err := buildInventoryChangedEvent("var-1")
	if err != nil {
		t.Fatal(err)
	}
	if event.Subscription != "InventoryChanged" {
		t.Fatalf("unexpected subscription: %s", event.Subscription)
	}
}

func TestVerifyGraphQLProvider(t *testing.T) {
	server, err := startProviderServer()
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() {
		_ = server.close(context.Background())
	})

	verifier := provider.NewVerifier()

	httpPact := filepath.Join("pacts", "product-consumer-product-provider.json")
	messagePact := filepath.Join("pacts", "messages", "product-consumer-product-provider.json")

	if err := verifier.VerifyProvider(t, provider.VerifyRequest{
		ProviderBaseURL: server.url,
		Provider:        "product-provider",
		PactFiles:       []string{httpPact},
	}); err != nil {
		t.Fatal(err)
	}

	messageHandlers := message.Handlers{
		"an inventory change message": func([]models.ProviderState) (message.Body, message.Metadata, error) {
			event, err := buildInventoryChangedEvent("var-1")
			if err != nil {
				return nil, nil, err
			}
			return event, message.Metadata{"contentType": "application/json"}, nil
		},
	}

	if err := verifier.VerifyProvider(t, provider.VerifyRequest{
		Provider:        "product-provider",
		PactFiles:       []string{messagePact},
		MessageHandlers: messageHandlers,
	}); err != nil {
		t.Fatal(err)
	}
}
